ui/app/src/components/DatasourceList/DatasourceEditorForm.tsx

```diff
@@ -0,0 +1,231 @@
+// Copyright 2023 The Perses Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { useImmer } from 'use-immer';
+import { Datasource, Display } from '@perses-dev/core';
+import {
+  Box,
+  Button,
+  Divider,
+  FormControl,
+  FormHelperText,
+  Grid,
+  InputLabel,
+  MenuItem,
+  Select,
+  Stack,
+  TextField,
+  Typography,
+} from '@mui/material';
+import { DispatchWithoutAction, useMemo } from 'react';
+import { PluginEditor } from '@perses-dev/plugin-system';
+import { useIsReadonly } from '../../model/config-client';
+
+// TODO: Replace with proper validation library
+function getValidation(state: Datasource) {
+  /** Name validation */
+  let name = null;
+  if (!state.metadata.name) {
+    name = 'Name is required';
+  }
+  // name can only contain alphanumeric characters and underscores and no spaces
+  if (state.metadata.name && !/^[a-zA-Z0-9_-]+$/.test(state.metadata.name)) {
+    name = 'Name can only contain alphanumeric characters, underscores, and dashes';
+  }
+
+  return {
+    name,
+    isValid: !name,
+  };
+}
+
+// this preprocessing ensure that we always have a defined object for the `display` property:
+function getInitialState(datasource: Datasource): Datasource {
+  const patchedDisplay: Display = {
+    name: '',
+    description: '',
+  };
+
+  if (datasource.spec.display) {
+    patchedDisplay.name = datasource.spec.display.name;
+    if (datasource.spec.display.description) {
+      patchedDisplay.description = datasource.spec.display.description;
+    }
+  }
+
+  return {
+    ...datasource,
+    spec: {
+      ...datasource.spec,
+      display: patchedDisplay,
+    },
+  };
+}
+
+interface DatasourceEditorFormProps {
+  initialDatasource: Datasource;
+  saveAction: string;
+  onSave: (datasource: Datasource) => void;
+  onCancel: DispatchWithoutAction;
+  onDelete: DispatchWithoutAction | undefined;
+  flagAsDraft: DispatchWithoutAction;
+}
+
+export function DatasourceEditorForm(props: DatasourceEditorFormProps) {
```

Comment:
Yes this discard modal is pretty much the same as for variables, I just added this flagAsDraft based stuff to avoid triggering it when nothing was changed (since on my side there is no read-specific mode, the case could occur quite often)